typescript/agentkit/src/wallet-providers/smartWalletProvider.ts

```diff
@@ -0,0 +1,372 @@
+   *   - Configuration parameters for setting up the smart wallet.
+   *
+   * @returns {Promise<SmartWalletProvider>}
+   *   - A promise that resolves to an instance of `SmartWalletProvider` configured with the provided settings.
+   *
+   * @throws {Error}
+   *   - If an invalid combination of `networkId` and `chainId` is provided.
+   *   - If the `chainId` cannot be determined.
+   *   - If the `chainId` is not supported.
+   *   - If `CDP_API_KEY_NAME` or `CDP_API_KEY_PRIVATE_KEY` is missing.
+   *
+   * @example
+   * ```typescript
+   * const smartWalletProvider = await SmartWalletProvider.configureWithWallet({
+   *   networkId: "base-sepolia",
+   *   signer: privateKeyToAccount("0xethprivatekey"),
+   *   cdpApiKeyName: "my-api-key",
+   *   cdpApiKeyPrivateKey: "my-private-key",
+   *   smartWalletAddress: "0x123456...",
+   * });
+   * ```
+   */
+  public static async configureWithSigner(
+    config: ConfigureSmartWalletWithSignerOptions,
+  ): Promise<SmartWalletProvider> {
+    const networkId = config.networkId || process.env.NETWORK_ID || Coinbase.networks.BaseSepolia;
+    const network = {
+      protocolFamily: "evm" as const,
+      chainId: NETWORK_ID_TO_CHAIN_ID[networkId],
+      networkId: networkId,
+    };
+
+    if (config.chainId) {
+      if (network.chainId && network.chainId != config.chainId) {
+        throw new Error(`Passed in a invalid combination of networkId and chainId`);
+      }
+      network.chainId = config.chainId;
+    }
+
+    if (!network.chainId) {
+      throw new Error(`Unable to determine chainId for network ${networkId}`);
+    }
+
+    const supportedChainIds = Object.keys(CHAIN_ID_TO_NETWORK_ID);
+    if (!supportedChainIds.includes(network.chainId)) {
+      throw new Error(
+        `Invalid chain id ${network.chainId}. Chain id must be one of ${supportedChainIds.join(", ")}`,
+      );
+    }
+
+    const cdpApiKeyName = config.cdpApiKeyName || process.env.CDP_API_KEY_NAME;
+    const cdpApiKeyPrivateKey = (
+      config.cdpApiKeyPrivateKey || process.env.CDP_API_KEY_PRIVATE_KEY
+    )?.replace(/\\n/g, "\n");
+
+    if (!cdpApiKeyName || !cdpApiKeyPrivateKey) {
+      throw new Error(
+        `CDP_API_KEY_NAME and CDP_API_KEY_PRIVATE_KEY are both required in SmartWalletProvider`,
+      );
+    }
+
+    Coinbase.configure({
+      apiKeyName: cdpApiKeyName as string,
+      privateKey: cdpApiKeyPrivateKey as string,
+    });
+
+    const smartWallet = config.smartWalletAddress
+      ? toSmartWallet({
+          signer: config.signer,
+          smartWalletAddress: config.smartWalletAddress,
+        })
+      : await createSmartWallet({
+          signer: config.signer,
+        });
+
+    const networkScopedSmartWallet = smartWallet.useNetwork({
+      chainId: Number(network.chainId) as SupportedChainId,
+    });
+
+    const smartWalletProvider = new SmartWalletProvider({
+      smartWallet: networkScopedSmartWallet,
+      network,
+      chainId: network.chainId,
```

Comment:
nit: Seems we can consolidate params – have the constructor accept just `network`, and then pick off `network.chainId` where needed. In fact, it looks like the current iteration of the constructor is not using `chainId`, so we can probably omit this entirely

Reply:
Yeah understood. I replied in another comment, but I was trying to think too much from a "chainId override" perspective when I started writing, but that was irrelevant because we're confined to the Smart Wallet supported networks regardless.